File: frontend/src/features/dynamic-simulation/algorithms/advanced_sw_ucb.ts
/**
 * Advanced Sliding Window UCB (Adv. SW-UCB) — TypeScript port
 * Reference: Garivier & Moulines (2011), with recovery detection modifications
 *
 * Enhanced SW-UCB with configurable exploration budget to detect when
 * previously failed gateways recover. Balances exploitation with proactive recovery detection.
 */

import { BaseAlgorithm, SeededRNG } from './base';
import { TransactionContext, ArmState, AlgorithmMetadata, HyperparameterSchema } from '../types';

export class AdvancedSlidingWindowUCB implements BaseAlgorithm {
    private arms: string[] = [];
    private windowSize = 200;
    private explorationRate = 0.05; // 5% default
    private history: Map<string, number[]> = new Map();
    private totalSelections = 0;
    private lastScores: Record<string, number> = {};
    private lastChosen = '';
    private lastWasExploration = false;
    private rng!: SeededRNG;

    initialize(arms: string[], config: Record<string, number>): void {
        this.arms = arms;
        this.windowSize = config.window_size ?? 200;
        this.explorationRate = config.exploration_rate ?? 0.05;
        this.history = new Map();
        for (const arm of arms) {
            this.history.set(arm, []);
        }
        this.totalSelections = 0;
        this.lastScores = {};
        this.lastChosen = '';
        this.lastWasExploration = false;
        this.rng = new SeededRNG(config.seed ?? 42);
    }

    select(_context: TransactionContext): string {
        this.totalSelections++;

        // Decide: explore or exploit?
        const useExploration = this.rng.random() < this.explorationRate && this.arms.length > 1;

        if (useExploration) {
            // Exploration: randomly select an arm
            this.lastWasExploration = true;
            const randomIndex = Math.floor(this.rng.random() * this.arms.length);
            this.lastChosen = this.arms[randomIndex];
            // Still compute scores for state tracking
            this.computeScores();
            return this.lastChosen;
        }

        // Exploitation: use UCB-based selection
        this.lastWasExploration = false;
        const scores: Record<string, number> = {};

        for (const arm of this.arms) {
            const hist = this.history.get(arm)!;
            const n = hist.length;
            if (n === 0) {
                this.lastChosen = arm;
                this.lastScores = Object.fromEntries(
                    this.arms.map(a => [a, a === arm ? Infinity : 0])
                );
                return arm;
            }
            const sr = hist.reduce((s, v) => s + v, 0) / n;
            const bonus = Math.sqrt(2 * Math.log(this.totalSelections) / n);
            scores[arm] = sr + bonus;
        }

        this.lastScores = scores;
        this.lastChosen = Object.entries(scores).reduce((best, [arm, score]) =>
            score > best[1] ? [arm, score] : best,
            ['', -Infinity]
        )[0] as string;
        return this.lastChosen;
    }

    private computeScores(): void {
        const scores: Record<string, number> = {};
        for (const arm of this.arms) {
            const hist = this.history.get(arm)!;
            const n = hist.length;
            if (n === 0) {
                scores[arm] = Infinity;
            } else {
                const sr = hist.reduce((s, v) => s + v, 0) / n;
                const bonus = Math.sqrt(2 * Math.log(this.totalSelections) / n);
                scores[arm] = sr + bonus;
            }
        }
        this.lastScores = scores;
    }

    update(arm: string, reward: number, _context: TransactionContext): void {
        const hist = this.history.get(arm)!;
        hist.push(reward);
        if (hist.length > this.windowSize) {
            hist.shift();
        }
    }

    getState(): Record<string, ArmState> {
        const state: Record<string, ArmState> = {};
        for (const arm of this.arms) {
            const hist = this.history.get(arm)!;
            const n = hist.length;
            const sum = hist.reduce((s, v) => s + v, 0);
            state[arm] = {
                estimatedSR: n > 0 ? sum / n : null,
                selectionScore: this.lastScores[arm] ?? null,
                totalSelections: n,
                windowCapacity: this.windowSize,
            };
        }
        return state;
    }

    explainLastDecision(): string {
        if (!this.lastChosen) return 'No decision made yet.';

        const hist = this.history.get(this.lastChosen)!;
        const n = hist.length;
        const sr = n > 0 ? hist.reduce((s, v) => s + v, 0) / n : 0;

        if (this.lastWasExploration) {
            return (
                `[EXPLORATION] Randomly chose '${this.lastChosen}' ` +
                `(SR=${sr.toFixed(3)}, window_n=${n}) to detect potential recovery.`
            );
        }

        const score = this.lastScores[this.lastChosen] ?? 0;
        return (
            `[EXPLOITATION] Chose '${this.lastChosen}' with UCB=${score.toFixed(4)} ` +
            `(SR=${sr.toFixed(3)}, window_n=${n})`
        );
    }

    getHyperparameterSchema(): HyperparameterSchema {
        return {
            window_size: {
                type: 'integer',
                default: 200,
                min: 10,
                max: 10000,
                description: 'Number of most recent transactions per gateway. Smaller = faster adaptation but noisier.',
            },
            exploration_rate: {
                type: 'number',
                default: 0.05,
                min: 0.0,
                max: 0.50,
                step: 0.01,
                description: 'Fraction of traffic for random exploration (0-50%). Higher = more probing for recovery.',
            },
        };
    }

    metadata(): AlgorithmMetadata {
        return {
            name: 'Advanced Sliding Window UCB',
            shortName: 'Adv. SW-UCB',
            description:
                'Enhanced SW-UCB with configurable exploration budget to detect when previously failed gateways recover. ' +
                'Balances exploitation with proactive recovery detection.',
            paper: 'Garivier & Moulines (2011), with recovery detection modifications',
            paperUrl: 'https://arxiv.org/abs/0805.3415',
            category: 'bandit',
        };
    }
}

File: algorithms/advanced_sw_ucb.py
"""
Advanced Sliding Window Upper Confidence Bound (Advanced SW-UCB) algorithm.
Reference: Garivier & Moulines (2011), with modifications for recovery detection.

Enhanced version of SW-UCB with dedicated exploration budget to detect
when previously failed gateways recover.
"""

import math
import random
from collections import deque
from typing import List, Dict, Any
from algorithms.base import BaseAlgorithm, TransactionContext


class AdvancedSlidingWindowUCB(BaseAlgorithm):
    """
    Enhanced SW-UCB with configurable exploration percentage.
    
    Balances exploitation (UCB-based) with exploration (random arm selection)
    to detect recovery in previously failing gateways. The exploration budget
    is dedicated to uniformly sampling non-best arms, enabling early detection
    of state changes in the environment.
    """

    def initialize(self, arms: List[str], config: Dict[str, Any]) -> None:
        self.arms = arms
        self.window_size = config.get("window_size", 200)
        self.exploration_rate = config.get("exploration_rate", 0.05)  # 5% default
        self.history: Dict[str, deque] = {
            arm: deque(maxlen=self.window_size) for arm in arms
        }
        self.total_selections = 0
        self._last_scores: Dict[str, float] = {}
        self._last_chosen: str = ""
        self._last_was_exploration: bool = False
        self.random_seed = config.get("seed", 42)
        random.seed(self.random_seed)

    def select(self, context: TransactionContext) -> str:
        self.total_selections += 1

        # Decide: explore or exploit?
        use_exploration = random.random() < self.exploration_rate

        if use_exploration and len(self.arms) > 1:
            # Exploration: randomly select an arm (uniform distribution)
            self._last_chosen = random.choice(self.arms)
            self._last_was_exploration = True
            # Still compute scores for state tracking
            self._compute_scores()
            return self._last_chosen

        # Exploitation: use UCB-based selection
        self._last_was_exploration = False
        scores = {}

        for arm in self.arms:
            hist = self.history[arm]
            n = len(hist)

            if n == 0:
                self._last_chosen = arm
                self._last_scores = {a: float('inf') if a == arm else 0.0 for a in self.arms}
                return arm

            sr = sum(hist) / n
            exploration_bonus = math.sqrt(2 * math.log(self.total_selections) / n)
            scores[arm] = sr + exploration_bonus

        self._last_scores = scores
        self._last_chosen = max(scores, key=scores.get)
        return self._last_chosen

    def _compute_scores(self) -> None:
        """Compute UCB scores for all arms (for state tracking even during exploration)."""
        scores = {}
        for arm in self.arms:
            hist = self.history[arm]
            n = len(hist)
            if n == 0:
                scores[arm] = float('inf')
            else:
                sr = sum(hist) / n
                exploration_bonus = math.sqrt(2 * math.log(self.total_selections) / n)
                scores[arm] = sr + exploration_bonus
        self._last_scores = scores

    def update(self, arm: str, reward: int, context: TransactionContext) -> None:
        self.history[arm].append(reward)

    def get_state(self) -> Dict[str, Any]:
        state = {}
        for arm in self.arms:
            hist = self.history[arm]
            n = len(hist)
            state[arm] = {
                "estimated_sr": sum(hist) / n if n > 0 else None,
                "selection_score": self._last_scores.get(arm),
                "window_count": n,
                "window_capacity": self.window_size,
                "window_successes": sum(hist),
                "window_failures": n - sum(hist),
            }
        return state

    def explain_last_decision(self) -> str:
        if not self._last_chosen:
            return "No decision made yet."

        if self._last_was_exploration:
            chosen = self._last_chosen
            hist = self.history[chosen]
            n = len(hist)
            sr = sum(hist) / n if n > 0 else 0
            others = {a: f"{s:.4f}" for a, s in self._last_scores.items() if a != chosen}
            return (
                f"[EXPLORATION] Randomly chose '{chosen}' "
                f"(SR={sr:.3f}, window_n={n}). "
                f"UCB scores: {others}"
            )

        # Exploitation path
        chosen = self._last_chosen
        score = self._last_scores.get(chosen, 0)
        hist = self.history[chosen]
        n = len(hist)
        sr = sum(hist) / n if n > 0 else 0
        bonus = score - sr if score != float('inf') else float('inf')
        others = {a: f"{s:.4f}" for a, s in self._last_scores.items() if a != chosen}
        return (
            f"[EXPLOITATION] Chose '{chosen}' with UCB={score:.4f} "
            f"(SR={sr:.3f} + bonus={bonus:.4f}, window_n={n}). "
            f"Other scores: {others}"
        )

    def get_hyperparameter_schema(self) -> Dict[str, Any]:
        return {
            "window_size": {
                "type": "integer",
                "default": 200,
                "min": 10,
                "max": 10000,
                "description": "Number of most recent transactions per gateway to consider. Smaller = faster adaptation but noisier.",
            },
            "exploration_rate": {
                "type": "number",
                "default": 0.05,
                "min": 0.0,
                "max": 0.50,
                "step": 0.01,
                "description": "Fraction of traffic dedicated to random exploration (0.0-0.50). Higher = more probing for recovery, lower = more stable. Default 0.05 = 5% exploration.",
            },
        }

    @classmethod
    def metadata(cls) -> Dict[str, str]:
        return {
            "name": "Advanced Sliding Window UCB",
            "short_name": "Adv. SW-UCB",
            "description": "Enhanced SW-UCB with configurable exploration budget to detect when previously failed gateways recover. Balances exploitation with proactive recovery detection.",
            "paper": "Garivier & Moulines (2011), with recovery detection modifications",
            "paper_url": "https://arxiv.org/abs/0805.3415",
            "category": "bandit",
            "non_stationary": "true",
        }
